---
title: "Pathway fingerprint: a tool for biomarker discovery based on gene expression data and pathway knowledge in R"
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{PFP}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
  \usepackage[utf8]{inputenc}
---

```{r, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.align = "center", 
  fig.show = "asis",
  eval = TRUE,
  tidy.opts = list(blank = FALSE, width.cutoff = 60),
  tidy = TRUE,
  message = FALSE,
  warning = FALSE
)
```

## Introduction

Recent network fingerprint method based on GO knowledge and propagation clustering mentioned provide a novel representation of network differentiation known as biological spectra, or Network Fingerprint. This method was used to describe the relationship between multiple disease networks and its related pathways, and to visually compare and parse different diseases by generating a fingerprint overlay. To give a quantitative comparison of the mixed disease network, we achieve the function of complex network comparison based on pathway on the open scientific computing platform **R**, and present *PFP* package for fingerprint-based pathway analyzing and comparison of systems. Driven by the research needs of customers, *PFP* provides a unified interface to three similarity clustering algorithms. In addition, *PFP* can also provide multiscale statistical analysis and visualization, access to the specific attributes of different disease fingerprint.
This manual is a brief introduction to structure, functions and usage of *PFP* package. The PFP package provides a set of functions to support knowledge-based network fingerprint (PFP) framework.
A biomedical network is characterized as a spectrum-like vector called “network fingerprint”, which contains similarities to basic reference networks. This framework provides a more intuitive way to decipher molecular networks, especially for large-scale network comparisons and clustering analyses.

The three main features of *PFP*:

- Basic reference networks generation.
- Network comparison, which encompasses network merging, annotation and
similarity scoring.
- Network standardization.

### Installation

*PFP* requires these packages: *magrittr*, *igraph*, *plyr*, *ggplot2*, *apcluster*, *dplyr*, *stringr*, *graph* and *KEGGgraph*. To install *PFP*, please note especially two depencies of *PFP*, *graph* and *KEGGgraph* are only available from [Bioconductor](http://bioconductor.org).
It also allows users to install the latest development version from github, which requires  *devtools* package has been installed on your system or can be installed using `install.packages("devtools")`. Note that devtools sometimes needs some extra non-R software on your system -- more specifically, an Rtools download for Windows or Xcode for OS X. There's more information about devtools [here](https://github.com/hadley/devtools).

```{r install-pkg-github, eval=FALSE}
## install PFP from github, require biocondutor dependencies package pre-installed 
if (!require(devtools))
  install.packages("devtools")
devtools::install_github("aib-group/PFP")
```
You  can also install *PFP* via Bioconductor.
```{r install-pkg-bioconductor, eval=FALSE}
## install PFP from github, require biocondutor dependencies package pre-installed 
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("PFP")
```
During analysis, you need to install *org.Hs.eg.db*, the installation strategy is as follows.
```{r install-database-bioconductor, eval=FALSE}
## install PFP from github, require biocondutor dependencies package pre-installed 
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("org.Hs.eg.db")
```

After installation, the \*{PFP} is ready to load into the current workspace by the following codes to the current workspace by typing or pasting the following codes:

```{r load-pkg,eval=TRUE, include=TRUE}
library(PFP)
```

## Analysis Pipeline: from Basic Reference Network Generation to Network
Networkfingerprint Visualization

We will demonstrate go through an analysis pipeline to illustrate some of the main functions in *PFP*. This pipeline consists of several steps:

1. Basic Reference Network Generation: prepare the well-known biomedical netowks as the PFP framework reference networks. Several pathway databases have been developed for biological network research, e.g. KEGG, Reactome - ([https://reactome.org](https://reactome.org)). All of this pathway databases is well-stuied and can be used as the basic reference networks of PFP.
1. Network fingerprint calculation: The similarity between two biomedical networks is calculated based on the following intition: grouping the nodes in the merged network into strongly inter-connected communities with high functional similarity score between intra-community nodes in different networks The functional similarity was measured based on GO [@ashburner2000gene]. And we employed affinity propagation (AP) [@frey2007clustering] clustering algorithm to detect the aligned functional modules between the two networks to be compared.
1. Network fingerprint Visualization: Show the network fingerprint along all the reference networks. We could observe the differences among biological networks fingerprint intuitively from visualization.

### Generating well-studied basic reference networks

The basic idea of calculating the network fingerprint is to have the biomedical networks map to well-studied basic networks. KEGG PATHWAY is a collection of manually drawn pathway maps representing our knowledge on the molecular interaction and reaction networks. Since its first introduction in 1995, KEGG PATHWAY has been widely used as a reference knowledge base for understanding biological pathways and functions of cellular processes. The knowledge from KEGG has proven of great value by numerous work in a wide range of fields [@kanehisa2007kegg]. So by default, we take KEGG pathways as basic reference networks in *PFP* by default.

Function `load_KEGG_refnet()` can be used to retrieve the KEGG pathway maps with KEGG API [http://www.kegg.jp/kegg/rest/keggapi.html](http://www.kegg.jp/kegg/rest/keggapi.html). In KEGG pathways, only the pathways of the map are manual drawing, and to different organisms, the map reference helps the automatic generation of organism-specific pathway for each organism. The `organism` (e.g. `organsim ="hsa"`) parameter indicate the organism name of KEGG pathway maps.


We defined a new S4 class `PFPrefnet` to store the *PFP* reference networks. *PFP* also provides five kinds of methods for this S4 class:

1. `net()`: Exact the basic reference networks of `PFPRefnet`.
1. `group()`: Obtain the group information, group names, number and the
size of each group,e.g. KEGG pathway database contains seven group pathway maps.
1. `subnet()`: Extract or replace parts of the *PFP* basic reference
networks.
1. `show()`: Display of `PFPRefnet`.
1. `name()`: Extract the names of reference networks.

Detailed instructions for this five methods refer to package function help.

Obviously users can also customize a PFPRefnet as a reference for computing network fingerprint. Users can refer to the documents of `PFPRefnet`  about the composition details of this class. *graphite* [@sales2012g] allow users to build `graphNEL` object from several pathway databases.


### Network fingerprint calculation

PFP algorithm consists of three steps: merging network, nodes clustering and similarity scoring.

Network merging. The two networks to be compared are first merged into one. Given two networks *G1* and *G2*, the merged network *Gm* is constructed by connecting each node between the *G1* and *G2* network. Two nodes corresponding to the same protein in the merged network are replaced by a single node that inherited all the interactions from the two individual nodes in the subsequent process.

Clustering in merged network. Grouping the nodes in the merged network into strongly inter-connected communities with high functional similarity score between intra-community nodes in different networks. We employed affinity propagation (AP) clustering algorithm to detect the aligned functional modules between the two networks to be compared. The nodes are grouped on the cluster based on nearest neighbor analysis.

Similarity scoring. The calculation of similarity score is processed in two steps: First, local similarity for each cluster and network similarity among cluster. Second, standardization: the original similarity score depends on the topological properties of query network to some extent. There is implicit bias of network fingerprint, because the outliers could be greatly distorted the relevant pattern presented in the network fingerprint. In order to eliminate the possible topological weight differences, the similarity calculation process of each node are standardized processing and the final network fingerprint facing to users is totally standardized. The standardization process is based on the random distribution of similarity scores. To the number of nodes, the number of edges and node degree, these three topological properties of random network for standardized estimate are consistent with the original network.

To not affect the results of standardization and improve the efficiency of fingerprint calculation, we set limit on the permutation time (the default is 100) of background network randomization in the standardization process. Users can also adjust randomization time of background network according to their own demands for the precision of network fingerprint.

*PFP* provides the `calc_sim_score()` function for calculating the similarity score of the network. As the similarity score is subjected to the size of the network, we use the maslov’s method [@maslov2002specificity] to randomize a network while preserving the degree distribution. The `nperm` parameter is added to `calc_sim_score()` refers to the permutation times (the default is 100) of random network while calculating the similarity score mentioned above. We define a S4 class `PFP` in our package to store the calculation results of network fingerprints.

Simply, we choose the two pathway maps `g` as a query network, and a subset networks of `kegg_refnet` as the reference networks. Then the PFP
can be calculated as following:
```{r a general-pipline,eval=TRUE, include=TRUE}
#load the data -- gene list of human; the PFPRefnet object of human; the PFP 
#object to test; the list of different genes.
data("gene_list_hsa")
data("PFPRefnet_hsa")
data("PFP_test1")
data("data_std")
# Step1: calculate the similarity score of network.
PFP_test <- calc_PFP_score(genes = gene_list_hsa,PFPRefnet = PFPRefnet_hsa)
# Step2: rank the pathway by the PFP score.
rank1 <- rank_PFP(object = PFP_test,total_rank = TRUE,thresh_value=0.5)
plot_PFP(rank1)
```
We study the target pathway, the pathway with the highest score after ranking.Below is 
a simple example.
```{r a the_target_gene,eval=TRUE, include=TRUE}
# Step1: select the max score of pathway.
pathway_select <- refnet_info(rank1)[1,"id"]
gene_test <- pathways_score(rank1)$genes_score[[pathway_select]]$ENTREZID
# Step2: get the correlation coefficient score of the edge.
edges_coexp <- get_exp_cor_edges(gene_test,data_std)
# Step3: Find the difference genes that are of focus.
gene_list2 <- unique(c(edges_coexp$source,edges_coexp$target))
# Step4: Find the edge to focus on.
edges_kegg <- get_bg_related_kegg(gene_list2,PFPRefnet=PFPRefnet_hsa,
                                  rm_duplicated = TRUE)
# Step5: Find the associated network
require(org.Hs.eg.db)
net_test <- get_asso_net(edges_coexp = edges_coexp,
                         edges_kegg = edges_kegg,
                         if_symbol = TRUE,
                         gene_info_db = org.Hs.eg.db)
```

            
### Network fingerprint visualization

*PFP* provides the `plot_PFP()` function to visualize the network fingerprint of a single query network.
First we show an example of PFP score.
```{r a PFP example, fig.height=6, fig.width=7.2, warning=FALSE}
plot_PFP(PFP_test)
```
Plot the scores from high to low.
```{r a rank PFP, fig.height=6, fig.width=7.2, warning=FALSE}
plot_PFP(rank1)
```

## Session Information

The version number of R and packages loaded for generating the vignette were:

```{r echo=FALSE}
sessionInfo()
```
